### DISCUSSION TITLE
What are rate limits for API?

### DISCUSSION THREAD

**Discussion by iBobik** [contember/admin#591] [Support] [answered]
Hi,
when I need to do some batch operations I can call Contember API calls in serial (await every call) or parallel (fire promises and the await them all). I found that about 5 parallel call are ok for Contember cloud, but in some features I could speed up some tasks a lot more if I can call e.g. 100 upsert mutations at one.

Is it ok to build features like this? Will your Cloud handle it all well (queue request as it needs) or there is risk of errors?

**Comment by osseonews** [comment]
This is a great question. Wondering as well, if their Cloud will be handle requests from a NextJs application, where thousands of pages, are generated statically at build time.

**Reply by iBobik** [reply]
Nuxt site with tens of thousands pages were ok.

But this kind of workers have some queue which allows e.g. 5-10 parallel requests.

**Reply by honzasladek** [reply]
Hi @osseonews, just want to confirm what @iBobik says. I've been using Contember Cloud with several Next.js projects deployed on Vercel with static site generation and it batches it so Cloud can easily handle that. (If you would hit it with thousands of large GraphQL queries at once, you would be stopped by rate limiting.)

**Reply by osseonews** [reply]
Thanks. What do you mean by: "it batches it." Where does the batching happen on Nextjs  or Contember Cloud?

**Reply by honzasladek** [reply]
In Next.js

**Comment by matej21** [comment, marked as answer]
Hello,

Let me provide some insights:

- **Hard limit on Nginx level**: Our infrastructure employs two nginx instances. Each instance is set up to allow 50 requests per second for your project, so in total, you're looking at 100 r/s. Importantly, nginx's `limit_req` directive is set with parameters `burst=200` and `delay=60`. This configuration allows for brief spikes in traffic – any excess requests beyond the rate limit will either be buffered and slowed down to fit within the allowed rate.

- **Handling Rate Limits**: When the rate limit is exceeded, the system will return a '429' status code. This indicates you've reached the rate limit and should adjust your request frequency.

- **Query Complexity Matters**: The number of requests is one thing, but the complexity of each query is another factor to consider. While you might be able to send multiple simple queries simultaneously, **more complex queries can be resource-intensive. It's recommended to run tests or benchmarks** for your specific queries to determine the optimal rate.

- **Database Serialization Failures**: When sending multiple mutations, there's a possibility of encountering **serialization failures** at the database level. Our system is built to handle these hiccups; it will attempt to process the mutation up to 5 times. While this ensures higher reliability, it might **increase the latency of some requests**.

- **Future Rate Limiting Plans**: We're looking into refining our rate limiting approach. Soon, we might implement engine-level rate limiting that's based not just on request count, but also on the actual complexity of the queries. This is something you'd want to keep in mind.

To sum it up: While you can certainly build features to use multiple parallel calls, it's essential to understand the underlying limits and mechanisms to ensure smooth operation. **Always monitor for '429' responses and adjust accordingly**.

Please be aware that this applies to our paid subscriptions and earlier free plans. The "developer sandbox" plan comes with significantly reduced limits. In the future, we anticipate differentiating the limits between the "hobby" and "pro" plans.

**Reply by iBobik** [reply]
Thank you for answer. So well written so you can copy&paste it in the docs. ;-)

**Comment by iBobik** [comment]
When I need to remove and create multiple entities with the same relation I can do it from mutation of the related entity (all in one request):

```
  updateProduct (
    by: { id: "01da456b-f8db-49f9-88dc-b483ed634b34" },
    data: { sizes: [
      { delete: { size: { id: "cecf9a30-b682-437c-bb57-17a6ae5992d2" } } },
      { create: { size: { connect: { id: "ef1f10cd-be41-489c-b050-1069c03200ac" } } } }
    ]}
  ) {
    ok
    errorMessage
  }
```

@matej21 Do you recommend to do it like this when possible, or it is better to remove and create entities one by one directly?

**Reply by matej21** [reply]
yes, this is correct usage.
